odoo/tools/float_utils.py

```diff
@@ -63,6 +65,19 @@ def float_round(value, precision_digits=None, precision_rounding=None, rounding_
     # In order to easily support rounding to arbitrary 'steps' (e.g. coin values),
     # we normalize the value before rounding it as an integer, and de-normalize
     # after rounding: e.g. float_round(1.3, precision_rounding=.5) == 1.5
+    def normalize(val):
+        return val / rounding_factor
+    def denormalize(val):
+        return val * rounding_factor
+
+    # inverting small rounding factors reduces rounding errors
+    if rounding_factor < 1:
+        invert_const = 1795630070.0  # accurately inverts up to {1,2,5}e-24
+        rounding_factor = invert_const / (invert_const * rounding_factor)
+        normalize, denormalize = denormalize, normalize
```

Comment:
I had some discussion with @odony and my first impression no longer holds.  We don't get really the reason for that magic number.  We would like some actual scientific explanation other than "it works".  By the way, wouldn't this work without the magic number, like this?
```suggestion
    # inverting small rounding factors reduces rounding errors
    if rounding_factor < 1:
        rounding_factor = 1 / rounding_factor
        normalize, denormalize = denormalize, normalize
```
Olivier came up with an even better idea.  We could go one step further, and use hardcoded conversions for the precision we _actually_ use.  We _never_ use rounding precisions like 1e-23, for instance.
```py
NORMALIZATION_FACTORS = {
  # prec:     factor
    .5:       2,
    .2:       5,
    .1:       10,
    .01:      100,
    .001:     1000,
    .0001:    10000,
    .00001:   100000,
    .000001:  1000000,
    .0000001: 10000000,
}
# (...)
    if rounding_factor < 1 and rounding_factor in NORMALIZATION_FACTORS:
        rounding_factor = NORMALIZATION_FACTORS[rounding_factor]
        normalize, denormalize = denormalize, normalize
```

Reply:
> We would like some actual scientific explanation other than "it works".

Here are the binary representations of common rounding factors & their inverted values: https://gist.github.com/lvsz/8bab40678c218722130fbab61c5a50c4
As you may notice, there isn't much of a pattern to their binary representations.

You can either try to multiply/divide them manually or use this calculator: http://weitz.de/ieee/

It "works" because `1795630070.0`s bits align in such a way that you can invert between them "losslessly" without rounding errors, as floating point operations basically cut off half the fractional part to keep everything the same precision. I doubt there's anything more to it than a happy coincidence as a consequence of representing decimal numbers using a binary system.

> By the way, wouldn't this work without the magic number, like this?

As I had shown in its [example usage](https://github.com/odoo/odoo/pull/152709#issuecomment-1991592127): no, you introduce rounding errors that way, which is what this PR aims to eliminate.

> Olivier came up with an even better idea. We could go one step further, and use hardcoded conversions for the precision we actually use. We never use rounding precisions like 1e-23, for instance.

Yes, I also thought of this, but it doesn't offer any additional accuracy or efficiency, and what if someone were to use a precision that isn't listed? E.g. 0.05 is useful for cash payments in Euro, do we throw an error? Make a ticket for bugfix to add it? Do we invert it naively and return a possible wrong result? Do we keep the magic number but only use it for factors that aren't hard-coded? In what way is this better?

Reply:
> use hardcoded conversions for the precision we actually use.

I think I found a workable compromise, hardcoding the most common ones, dynamically adding others by inverting them as a `Decimal`:

```python
from collections import defaultdict
from decimal import Decimal


class invertdict(defaultdict):
    def __init__(self, *args, **kwargs):
        return super().__init__(lambda val: float(1 / Decimal(str(val))), *args, **kwargs)

    def __missing__(self, key):
        value = self[key] = self.default_factory(key)
        return value

NORMALIZATION_FACTORS = invertdict([
    (0.5, 2),
    (0.2, 5),
    (0.1, 10),
    (0.01, 100),
    (0.001, 1000),
    (0.0001, 10000),
])
```

Reply:
As the inversion method has to be replicable in JS, I had to find something that doesn't used `Decimal`:

https://github.com/odoo/odoo/blob/3f050e79e7ae93f8c81b2550df77f7887a79509f/odoo/tools/float_utils.py#L278-L283

It converts the number to scientific notation, then inverts the sign of the exponent, and inverts the coefficient by dividing by it's square ( ${x \over x^2} = {1 \over x}$ )

It's neither as fast nor as accurate as either the magic number or `Decimal`, but I should be fine as long as `precision_rounding` is larger than `1e-11` (compared to `1e-5` for the naive `def invert(val): return 1 / val` approach).